cirq/contrib/routing/greedy.py

```diff
@@ -0,0 +1,253 @@
+            physical_op = logical_op.with_qubits(
+                *self.log_to_phys(*logical_op.qubits))
+            assert len(physical_op.qubits
+                      ) < 2 or physical_op.qubits in self.device_graph.edges
+            self.physical_ops.append(physical_op)
+        return
+
+    @property
+    def swap_network(self):
+        return SwapNetwork(circuits.Circuit.from_ops(self.physical_ops),
+                           self.initial_mapping)
+
+    def distance(self, edge):
+        return self.physical_distances[tuple(self.log_to_phys(*edge))]
+
+    def swap_along_path(self, path):
+        for i in range(len(path) - 1):
+            self.apply_swap(path[i:i + 2])
+
+    def bring_farthest_pair_together(self,
+                                     pairs: Iterable[Tuple[ops.Qid, ops.Qid]]):
+        distances = [self.distance(pair) for pair in pairs]
+        assert distances
+        max_distance = min(distances)
+        farthest_pairs = [
+            pair for pair, d in zip(pairs, distances) if d == max_distance
+        ]
+        farthest_pair = random.choice(farthest_pairs)
+        edge = self.log_to_phys(*farthest_pair)
+        shortest_path = nx.shortest_path(self.device_graph, *edge)
+        assert len(shortest_path) - 1 == max_distance
+        midpoint = max_distance // 2
+        self.swap_along_path(shortest_path[:midpoint])
+        self.swap_along_path(shortest_path[midpoint:])
+
+    def get_distance_vector(self, logical_edges, swaps):
+        self.update_mapping(*swaps)
+        distance_vector = np.array([self.distance(e) for e in logical_edges])
+        self.update_mapping(*swaps)
+        return distance_vector
+
+    def apply_next_swaps(self):
+        timeslices = get_timeslices(self.remaining_dag)
+        for k in range(1, self.max_search_radius + 1):
+            candidate_swap_sets = list(self.get_edge_sets(k))
+            for timeslice in timeslices:
+                distance_vectors = list(
+                    self.get_distance_vector(timeslice.edges, swap_set)
+                    for swap_set in candidate_swap_sets)
+                dominated_indices = get_dominated_indices(distance_vectors)
+                candidate_swap_sets = [
+                    S for i, S in enumerate(candidate_swap_sets)
+                    if i not in dominated_indices
+                ]
+                if len(candidate_swap_sets) == 1:
+                    self.apply_swap(*candidate_swap_sets[0])
+                    return
+
+        frontier_edges = timeslices[0].edges
+        self.bring_farthest_pair_together(frontier_edges)
+
+    def route(self):
+        self.apply_possible_ops()
+        while self.remaining_dag:
+            self.apply_next_swaps()
+            self.apply_possible_ops()
+        assert get_ops_consistency_with_device_graph(self.physical_ops,
+                                                     self.device_graph)
+
+
+def get_dominated_indices(distance_vectors: List[np.ndarray]):
```

Comment:
docstring. should this function be private?

Reply:
Done